gcc/d/config-lang.in

```diff
@@ -25,7 +25,18 @@ language="d"
 
 compilers="d21\$(exeext)"
 
-target_libs="target-libphobos target-zlib target-libbacktrace"
+case "${noconfigdirs}" in
+  # Check if phobos was disabled as unsupported
+  *target-libphobos*) ;;
+  *)
+    # The --disable-<component> handler in configure.ac is called after
+    # config-lang.in. So when using --disable-libphobos, it has not been
+    # added to noconfigdirs here yet
+    if eval test x\${enable_${libphobos}} "!=" xno ; then
+      target_libs="target-libphobos target-zlib target-libbacktrace"
```

Comment:
You will also need to add `noconfigdirs=target-zlib` if libphobos is disabled, so that it won't be built as well.

Reply:
Thanks for reviewing. But where would enabling D have an effect on enabling `target-zlib`? AFAICS there's `Makefile.def` which makes `target-libphobos` depend on `target-zlib`, but if we disable `target-libphobos` in `configure.ac`/`config-lang.in` the `Makefile.def` dependencies should not have any effect. And I think in `configure.ac`/`config-lang.in` there's no code which should pull in `target-zlib`, except for the `target_libs` code which is already handled in this PR.

I just fear that if we explicitly add `noconfigdirs=target-zlib` we might break builds where GCC depends on `target-zlib` for some other reason?

Reply:
It's a change I've only managed to get in last night: `target-zlib` has been added back into top-level `target_libraries` here: https://github.com/gcc-mirror/gcc/commit/e60a24bda6d953531ffa7c02af296dad32603c2d#diff-67e997bcfdac55191033d57a16d1408a

This means that if the D front-end is not being built, it will be automatically added to `noconfigdirs`.   However if d/config-lang.in does not include target-zlib as a target_lib, then zlib will always be built, regardless of whether it's required.

Reply:
Ah, now I understand: In `configure.ac`, by default all libraries are enabled. `configure.ac` then sources the `config-lang.in` fragment and if the language is unsupported / disabled, it adds `target_libs` to `disabled_target_libs` and later to `noconfigdirs`: https://github.com/gcc-mirror/gcc/blob/e60a24bda6d953531ffa7c02af296dad32603c2d/configure.ac#L2026 .

But this should also be a problem for `target-libphobos` and `target-libbacktrace`, I think? I suppose the best solution here is to add all libs to `disabled_target_libs` instead in all cases where we do not set `target_libs`: `disabled_target_libs="$disabled_target_libs target-libphobos target-zlib target-libbacktrace"`.

Reply:
> But this should also be a problem for target-libphobos and target-libbacktrace, I think?

For target-libphobos, I think no, as if it's explicitly disabled, or unsupported, then the top-level configure will add it to `noconfigdirs`.  I don't think it will be pull in automatically, but I'll have to check that. :-)

target-libbacktrace is used by other front-ends, so leave it alone.

Reply:
Well, 
* here all known target libraries are added to `target_configdirs`: https://github.com/gcc-mirror/gcc/blob/e60a24bda6d953531ffa7c02af296dad32603c2d/configure.ac#L185
* explicit `--disable-X` adds `X` to `noconfigdirs`: https://github.com/gcc-mirror/gcc/blob/e60a24bda6d953531ffa7c02af296dad32603c2d/configure.ac#L2123
* Here, `noconfigdirs` are removed from `target_configdirs`: https://github.com/gcc-mirror/gcc/blob/e60a24bda6d953531ffa7c02af296dad32603c2d/configure.ac#L2177
* `target_configdirs` are added to the makefile targets: https://github.com/gcc-mirror/gcc/blob/e60a24bda6d953531ffa7c02af296dad32603c2d/configure.ac#L2878

So I think if we do nothing, all target libs including `target-libphobos` will end up in the Makefile targets.

Disabling unused libraries happens as explined in my previous post: The `config-lang.in` fragment is invoked for all languages, even disabled ones. Then if the language is disabled, it adds `target_libs` to `disabled_target_libs`, otherwise to  `enabled_target_libs` https://github.com/gcc-mirror/gcc/blob/e60a24bda6d953531ffa7c02af296dad32603c2d/configure.ac#L2026 . After this has been done for all languages, all libs which are in `disabled_target_libs` **and not in** `enabled_target_libs` are added to `noconfigdirs` https://github.com/gcc-mirror/gcc/blob/e60a24bda6d953531ffa7c02af296dad32603c2d/configure.ac#L2060.

This means we should not modify `noconfigdirs`, as it will mess up other languages. We should however adjust `disabled_target_libs` in the same way it would be done if the `d` language is not enabled. Even if we add `libbacktrace` here, it will not be disabled if some other frontend put it in `target_libs`, as this will put it into `enabled_target_libs`.

Reply:
Applying this patch and forcing libphobos to be unsupported.
* --enable-languages=c: `noconfigdirs=target-libphobos target-libbacktrace`
* --enable-languages=d: `noconfigdirs=target-libphobos target-libbacktrace`
* --enable-languages=d --disable-libphobos: `noconfigdirs=target-libphobos`
* --enable-languages=c --enable-libphobos: `noconfigdirs=target-libphobos target-zlib target-libbacktrace`

Without this patch (only forcing unsupported).
* --enable-languages=c: `noconfigdirs=target-libphobos target-zlib target-libbacktrace`
* --enable-languages=d: `noconfigdirs=target-libphobos`
* --enable-languages=d --disable-libphobos: `noconfigdirs=target-libphobos`
* --enable-languages=c --enable-libphobos: `noconfigdirs=target-libphobos target-zlib target-libbacktrace`

Reply:
Seems mostly correct:
* `libphobos` is always in there, as you marked it as unsupported

this patch,
* --enable-languages=c: It's missing the `target-zlib` compared to the variant without patch. This is because we do not set `target_libs` and therefore `disabled_target_libs` are not set. libbacktrace is disabled as it is in fortrans `target_libs` and we do not build fortran.
* --enable-languages=d: Same as above.
* --enable-languages=d --disable-libphobos: This is strange. Likely a problem in the patch, I'll have a look.
* --enable-languages=c --enable-libphobos: I don't really get why this would disabled zlib but --enable-languages=c alone does not?

without this patch:
* --enable-languages=d: we want the same result as for --enable-languages=c. This shows the problem that we pull in backtrace/zlib even if phobos is not used.
* --enable-languages=c --enable-libphobos: same as --enable-languages=c, I guess that's expected.

I'll do some testing and push an updated patch.

Reply:
Setting `disabled_target_libraries`.
* --enable-languages=c: noconfigdirs=target-libphobos target-zlib target-libbacktrace
* --enable-languages=d: noconfigdirs=target-libphobos target-zlib target-libbacktrace
* --enable-languages=d --disable-libphobos: noconfigdirs=target-libphobos target-zlib target-libbacktrace
* --enable-languages=c --enable-libphobos: noconfigdirs=target-libphobos target-zlib target-libbacktrace


By the way, the condition should be:
```
if eval test x${enable_libphobos} "!=" xno ; then
```

Reply:
> By the way, the condition should be:
> if eval test x${enable_libphobos} "!=" xno ; then

Because I got some wrong results from the condition in some of the test vectors.

Reply:
@ibuclaw your're right, there was an error in that if condition. I've pushed a new patch and I can reproduce your results, I think this is working properly now and most cases should have been tested:

## no patch
supported

| Flags | noconfigdirs |
| --------|-----------------|
|--enable-languages=c| gotools target-libgo  target-libphobos target-zlib target-libbacktrace|
|--enable-languages=d| gotools target-libgo |
|--enable-languages=d --disable-libphobos| gotools target-libgo target-libphobos|
|--enable-languages=c --disable-libphobos| gotools target-libgo  target-libphobos target-zlib target-libbacktrace|
|--enable-languages=c --enable-libphobos| gotools target-libgo target-libphobos target-zlib target-libbacktrace |


unsupported

| Flags | noconfigdirs |
| --------|-----------------|
|--enable-languages=c|target-libphobos gotools target-libgo target-libphobos target-zlib target-libbacktrace |
|--enable-languages=d|target-libphobos gotools target-libgo |
|--enable-languages=d --disable-libphobos| gotools target-libgo target-libphobos|
|--enable-languages=c --disable-libphobos| gotools target-libgo target-libphobos target-zlib target-libbacktrace |
|--enable-languages=c --enable-libphobos| gotools target-libgo target-libphobos target-zlib target-libbacktrace |

## patch
supported

| Flags | noconfigdirs |
| --------|-----------------|
|--enable-languages=c| gotools target-libgo target-libphobos target-zlib target-libbacktrace |
|--enable-languages=d| gotools target-libgo |
|--enable-languages=c,go| target-libphobos target-zlib |
|-enable-languages=d,go --disable-libphobos|target-libphobos target-zlib|
|--enable-languages=d --disable-libphobos| gotools target-libgo target-libphobos target-zlib target-libbacktrace|
|--enable-languages=c --disable-libphobos| gotools target-libgo target-libphobos target-zlib target-libbacktrace target-libphobos|
|--enable-languages=c --enable-libphobos| gotools target-libgo target-libphobos target-zlib target-libbacktrace |

unsupported

| Flags | noconfigdirs |
| --------|-----------------|
|--enable-languages=c|target-libphobos gotools target-libgo target-libphobos target-zlib target-libbacktrace|
|--enable-languages=d|target-libphobos gotools target-libgo target-libphobos target-zlib target-libbacktrace |
|--enable-languages=c,go| target-libphobos target-zlib |
|--enable-languages=d --disable-libphobos| gotools target-libgo target-libphobos target-zlib target-libbacktrace target-libphobos|
|--enable-languages=c --disable-libphobos| gotools target-libgo target-libphobos target-zlib target-libbacktrace target-libphobos|
|--enable-languages=c --enable-libphobos| gotools target-libgo target-libphobos target-zlib target-libbacktrace |

Reply:
Seems reasonable, apart from target-libphobos being added twice I guess.

I wonder if it might be better to have this in the top-level configure script though.  I don't really have an opinion either way.

Reply:
Updated to not apply `target-libphobos` twice. Doing this in the top-level configure would be more complicated: Removing libs from `enabled_target_libs` is error prone. The other solution is to merge this into the code sourcing `config-lang.in`, but that's supposed to be generic code and we would have to special-case `d/config-lang.in` there.

The best solution would probably be to introduce a notion of dependecies for target-libs, then depend only on `target-libphobos`. This would require some larger changes in `configure.ac` though, and it's probably too late to do that for GCC9.

So can I post the patch to the ML? Is there anyone in particular I should CC?